File: whiteboard-client/src/canvas/__tests__/CanvasFrameBinaryConverter.test.ts
import { test, expect } from "vitest";
import CanvasFrameBinaryConverter from "../CanvasFrameBinaryConverter";

import type { CanvasFrame } from "../types";

const testImageData = new Uint8ClampedArray([
  1, 1, 1, 255, 2, 2, 2, 255, 3, 3, 3, 255, 4, 4, 4, 255,
]);

function createTestBinaryFrame() {
  const view = new DataView(new ArrayBuffer(4 * 4 + 16));
  let pos = 0;

  view.setUint32(pos, 1);
  pos += 4;

  view.setUint32(pos, 2);
  pos += 4;

  view.setUint32(pos, 3);
  pos += 4;

  view.setUint32(pos, 4);
  pos += 4;

  for (const b of testImageData) {
    view.setUint8(pos++, b);
  }

  return view;
}

test("canvas frame is converted from bytes", () => {
  const view = createTestBinaryFrame();
  const converter = new CanvasFrameBinaryConverter();

  const frame = converter.fromBytes(view.buffer) as CanvasFrame;

  expect(frame.startX).toEqual(1);
  expect(frame.startY).toEqual(2);
  expect(frame.endX).toEqual(3);
  expect(frame.endY).toEqual(4);

  for (let i = 0; i < testImageData.length; i++) {
    expect(frame.data[i]).toEqual(testImageData[i]);
  }
});

test("canvas frame is converted to bytes", () => {
  const view = createTestBinaryFrame();
  const converter = new CanvasFrameBinaryConverter();
  const frame = {
    startX: 1,
    startY: 2,
    endX: 3,
    endY: 4,
    data: testImageData,
  };

  const bytes = converter.toBytes(frame);

  const result = new DataView(bytes);
  let pos = 0;

  expect(result.getUint32(pos)).toEqual(1);
  pos += 4;

  expect(result.getUint32(pos)).toEqual(2);
  pos += 4;

  expect(result.getUint32(pos)).toEqual(3);
  pos += 4;

  expect(result.getUint32(pos)).toEqual(4);
  pos += 4;

  for (const b of testImageData) {
    expect(result.getUint8(pos++)).toEqual(b);
  }
});

File: server/src/test/java/dev/oscarrojas/kittypen/canvas/CanvasFrameBinaryConverterTest.java
package dev.oscarrojas.kittypen.canvas;

import dev.oscarrojas.kittypen.core.canvas.CanvasFrame;
import dev.oscarrojas.kittypen.core.canvas.CanvasFrameBinaryConverter;
import org.junit.jupiter.api.Test;

import java.nio.ByteBuffer;

import static org.junit.jupiter.api.Assertions.assertEquals;

class CanvasFrameBinaryConverterTest {

    @Test
    void fromBytes() {
        // 4 32-bit integers (4 * 4 bytes) for frame bounds and 16 bytes of frame data
        ByteBuffer buffer = ByteBuffer.allocate(4 * 4 + 16);
        // startX
        buffer.putInt(1);
        // startY
        buffer.putInt(2);
        // endX
        buffer.putInt(3);
        // endY
        buffer.putInt(4);
        // frame image data
        byte[] data = new byte[]{1, 1, 1, (byte) 255, 2, 2, 2, (byte) 255, 3, 3, 3, (byte) 255, 4, 4, 4, (byte) 255};
        for (byte b : data) {
            buffer.put(b);
        }

        CanvasFrameBinaryConverter decoder = new CanvasFrameBinaryConverter();

        CanvasFrame frame = decoder.fromBytes(buffer.array());

        assertEquals(1, frame.getStartX());
        assertEquals(2, frame.getStartY());
        assertEquals(3, frame.getEndX());
        assertEquals(4, frame.getEndY());
        for (int i = 0; i < data.length; i++) {
            assertEquals(data[i], frame.getData()[i]);
        }
    }

}